# Laboratorio 1 Aprendizaje de Máquina

Integrantes del grupo:
1. Emmanuel Blanco - 202312743
2. Juan David Guzmán - #

## Exploración de Datos

Para poder hacer la exploración de los datos primero importamos todas la librerías que nos pueden resultar útiles:

In [5]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
#TODO: Incluir importaciones necesarias de scikit-learn aquí
#from sklearn.preprocessing import 

A continuación incluimos las versiones de las librerías que empleamos:

In [6]:
from importlib.metadata import version
print(f"Versión de Pandas: {version('pandas')}")
print(f"Versión de Matplotlib: {version('matplotlib')}")
print(f"Versión de Seaborn: {version('seaborn')}")
print(f"Versión de Scikit learn: {version('scikit-learn')}")

Versión de Pandas: 2.2.3
Versión de Matplotlib: 3.10.0
Versión de Seaborn: 0.13.2
Versión de Scikit learn: 1.6.1


En el siguiente espacio para código desplegamos la información contenida en el diccionario de datos para informanos sobre lo que contienen los datos

In [12]:
#Cargamos en diccionario de datos
dict_datos = pd.read_excel("./data/DiccPacientes.xlsx")
pd.set_option('display.max_colwidth', None)
dict_datos

Nombre Columna  Tipo de dato  \
0                     Patient ID       String   
1                Date of Service         Date   
2                            Sex       String   
3                            Age      Integer   
4                    Weight (kg)        Float   
5                     Height (m)        Float   
6                            BMI        Float   
7   Abdominal Circumference (cm)        Float   
8          Blood Pressure (mmHg)       String   
9      Total Cholesterol (mg/dL)        Float   
10                   HDL (mg/dL)        Float   
11   Fasting Blood Sugar (mg/dL)        Float   
12                Smoking Status      Boolean   
13               Diabetes Status      Boolean   
14       Physical Activity Level       String   
15         Family History of CVD      Boolean   
16                   Height (cm)        Float   
17         Waist-to-Height Ratio        FLoat   
18                   Systolic BP      Integer   
19                  Diastolic BP      Integer   
20       Blood Pressure Category       String   
21         Estimated LDL (mg/dL)        Float   
22                CVD Risk Score        Float   
23                CVD Risk Level      String    

                                                                                              Comentarios  
0                                                                              Identificador del paciente  
1                                                                                    Fecha de la atención  
2                                                                              Sexo (Femenino, Masculino)  
3                                                                                                    Edad  
4                                                                                                    Peso  
5                                                                                                  Altura  
6                                                                                 Índice de masa corporal  
7                                                                                Circunferencia abdominal  
8             Presión sanguínea, de la forma "<Presión arterial sistólica>/<Presión arterial diastólica>"  
9                                                                                        Colesterol total  
10                                                 Lipoproteína de alta densidad en el torrente sanguíneo  
11                                                                  Nivel de azucar en la sangre en ayuno  
12                                                          Booleano que indica si el paciente es fumador  
13                                                        Booleano que indica si el paciente es diabético  
14                                                       Nivel de actividad física (Alto, Moderado, Bajo)  
15             Booleano que indica si en la familia han habido antecedentes de enfermedad cardio vascular  
16                                                                                               Estatura  
17                                             Cociente entre la circumferencia de la cintura y la altura  
18                                                                             Presión arterial sistólica  
19                                                                            Presión arterial diastólica  
20  Categoría de presión sanguínea (Normal, Elevada, Hipertensión Etapa 1, Elevada, Hipertensión Etapa 2)  
21                                                 Lipoproteína de baja densidad en el torrente sanguíneo  
22                                                                                  Riesgo cardiovascular  
23                                                 Nivel de riesgo cadiovascular (Alto, Intermedio, Bajo)

Ahora pasamos a cargar los datos de prueba dentro de un dataframe para poder revisarlos e identificar detalles que debamos corregir o preparar para luego pasárselos al modelo

In [13]:
test_data = pd.read_csv("./data/Datos_Test_Lab_1.csv",sep=";")
test_data

Patient ID   Date of Service Sex   Age  Weight (kg)  Height (m)     BMI  \
0     WYTG8678         1/12/2021   M  44.0       51.936       1.936  36.442   
1     yloc1993        12-25-2020   F  30.0      115.000       1.830  34.300   
2     DDwq7653         23-sep-21   M  38.0       63.300       1.700  21.900   
3     YTcX6704         3/03/2021   F  58.0       72.300       1.740  23.900   
4     lZLi8935        06-14-2022   M  51.0       57.600       1.810  17.600   
..         ...               ...  ..   ...          ...         ...     ...   
189   wTEx7709        14/05/2025   F  64.0       62.584       1.627  38.852   
190   YHlG8563  October 05, 2020   M  42.0       67.600       1.810  20.600   
191   Axab9332        11/12/2021   F  58.0       69.870       1.944  20.785   
192   mAlf8380          7-may-22   F  37.0       79.000       1.740     NaN   
193   BBiM3801         9/12/2021   F  64.0       67.144       1.532  22.811   

     Abdominal Circumference (cm) Blood Pressure (mmHg)  \
0                         115.755               105/108   
1                          93.600                139/79   
2                          93.800                 93/67   
3                          91.600                 99/90   
4                         100.000                 97/84   
..                            ...                   ...   
189                        92.159                143/84   
190                        98.300                113/72   
191                        86.259               162/106   
192                        92.400                147/85   
193                        99.790                127/99   

     Total Cholesterol (mg/dL)  ...  Smoking Status  Diabetes Status  \
0                        107.0  ...               Y                Y   
1                        100.0  ...               Y                N   
2                          NaN  ...               N                Y   
3                        208.0  ...               N                N   
4                        137.0  ...               Y                N   
..                         ...  ...             ...              ...   
189                      241.0  ...               Y                Y   
190                      258.0  ...               Y                Y   
191                      151.0  ...               Y                N   
192                      212.0  ...               Y                N   
193                      260.0  ...               N                N   

    Physical Activity Level Family History of CVD Height (cm)  \
0                       Low                     N     193.610   
1                  Moderate                     N     183.000   
2                      High                     N     170.000   
3                      High                     N     174.000   
4                      High                     N     181.000   
..                      ...                   ...         ...   
189                    High                     Y     162.655   
190                Moderate                     N     181.000   
191                Moderate                     N     194.450   
192                Moderate                     N     174.000   
193                    High                     N     153.171   

    Waist-to-Height Ratio  Systolic BP  Diastolic BP  Blood Pressure Category  \
0                   0.598        105.0         108.0     Hypertension Stage 2   
1                   0.511        139.0          79.0     Hypertension Stage 1   
2                   0.552         93.0          67.0                   Normal   
3                   0.526         99.0          90.0     Hypertension Stage 2   
4                   0.552         97.0          84.0     Hypertension Stage 1   
..                    ...          ...           ...                      ...   
189                 0.567        143.0          84.0     Hypertension Stage 1   
190                 0.543        113.0          72.0

Tamaño del dataframe:

In [25]:
print(test_data.shape)

(194, 22)


Tipo de los datos de las columnas de nuestro dataframe

In [14]:
test_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 194 entries, 0 to 193
Data columns (total 22 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Patient ID                    194 non-null    object 
 1   Date of Service               194 non-null    object 
 2   Sex                           194 non-null    object 
 3   Age                           183 non-null    float64
 4   Weight (kg)                   184 non-null    float64
 5   Height (m)                    187 non-null    float64
 6   BMI                           186 non-null    float64
 7   Abdominal Circumference (cm)  187 non-null    float64
 8   Blood Pressure (mmHg)         194 non-null    object 
 9   Total Cholesterol (mg/dL)     188 non-null    float64
 10  HDL (mg/dL)                   186 non-null    float64
 11  Fasting Blood Sugar (mg/dL)   189 non-null    float64
 12  Smoking Status                194 non-null    object 
 13  Diabe

Descripción de los datos

In [15]:
test_data.describe()

Age  Weight (kg)  Height (m)         BMI  \
count  183.000000   184.000000  187.000000  186.000000   
mean    46.464481    85.583049    1.756995   28.448113   
std     12.312220    22.159352    0.118687    7.024220   
min     25.000000    19.093000    1.398000    6.698000   
25%     38.000000    65.362250    1.660000   22.908250   
50%     44.000000    86.962500    1.760000   28.371500   
75%     55.000000   106.525000    1.850000   33.575000   
max     77.000000   119.700000    1.989000   49.649000   

       Abdominal Circumference (cm)  Total Cholesterol (mg/dL)  HDL (mg/dL)  \
count                    187.000000                  188.00000   186.000000   
mean                      93.666995                  196.68617    55.507581   
std                       12.413227                   60.02613    15.997907   
min                       70.900000                  100.00000     3.410000   
25%                       82.724000                  137.00000    42.250000   
50%                       95.209000                  197.00000    55.500000   
75%                      103.050000                  251.00000    67.000000   
max                      134.089000                  299.00000    89.000000   

       Fasting Blood Sugar (mg/dL)  Height (cm)  Waist-to-Height Ratio  \
count                   189.000000   187.000000             188.000000   
mean                    121.249915   175.748813               0.534793   
std                      31.239646    11.399690               0.074244   
min                      16.234000   150.391000               0.377000   
25%                      98.000000   166.000000               0.484750   
50%                     123.000000   176.000000               0.532000   
75%                     141.000000   185.000000               0.586500   
max                     197.000000   198.949000               0.790000   

       Systolic BP  Diastolic BP  Estimated LDL (mg/dL)  
count   188.000000    189.000000             189.000000  
mean    125.618463     83.301741             110.572058  
std      22.749990     14.753023              62.820276  
min      90.000000     60.000000              -6.000000  
25%     107.000000     70.000000              58.000000  
50%     121.000000     84.000000             106.000000  
75%     142.000000     93.000000             161.000000  
max     195.271000    130.029000             289.119000

Podemos ver que el conteo de algunas columnas resulta distinta de 194, lo que indica falta de datos en algunas columnas. Con base en eso ejecutamos la siguiente instrucción:

In [24]:
((test_data.isnull().sum()/test_data.shape[0])).sort_values(ascending=False)

Age                             0.056701
Weight (kg)                     0.051546
HDL (mg/dL)                     0.041237
BMI                             0.041237
Height (m)                      0.036082
Abdominal Circumference (cm)    0.036082
Height (cm)                     0.036082
Waist-to-Height Ratio           0.030928
Total Cholesterol (mg/dL)       0.030928
Systolic BP                     0.030928
Estimated LDL (mg/dL)           0.025773
Diastolic BP                    0.025773
Fasting Blood Sugar (mg/dL)     0.025773
Patient ID                      0.000000
Sex                             0.000000
Date of Service                 0.000000
Blood Pressure (mmHg)           0.000000
Smoking Status                  0.000000
Diabetes Status                 0.000000
Family History of CVD           0.000000
Physical Activity Level         0.000000
Blood Pressure Category         0.000000
dtype: float64

Hacen falta datos en las columnas Age, Weight, HDL, BMI, Height, Abdominal Circumference, Height (cm), Waist-to-Height ratio, Total Cholesterol, Systolic BP, Diastolic BP, Estimated LDL y Fasting Blood Sugar

In [39]:
# Filtrar filas con Calificación nula
nulos_age = test_data[test_data['Age'].isna()].copy()
nulos_weight = test_data[test_data["Weight (kg)"].isna()].copy()
nulos_HDL = test_data[test_data["HDL (mg/dL)"].isna()].copy()
nulos_BMI = test_data[test_data["BMI"].isna()].copy()
nulos_height_m = test_data[test_data["Height (m)"].isna()].copy()
nulos_abdmomin = test_data[test_data["Abdominal Circumference (cm)"].isna()].copy()
nulos_height_cm = test_data[test_data["Height (cm)"].isna()].copy()
nulos_w_h_ratio = test_data[test_data["Waist-to-Height Ratio"].isna()].copy()
nulos_colest = test_data[test_data["Total Cholesterol (mg/dL)"].isna()].copy()
nulos_systolic = test_data[test_data["Systolic BP"].isna()].copy()
nulos_blood_sugar = test_data[test_data["Fasting Blood Sugar (mg/dL)"].isna()].copy()
nulos_estimated_LDL = test_data[test_data["Estimated LDL (mg/dL)"].isna()].copy()

# Contar cuántas son
print(f"Filas con Age nula: {nulos_age.shape[0]}")
print(f"Filas afectadas: {test_data.index[test_data["Age"].isna()]}")
print(f"Filas con Weight nula: {nulos_weight.shape[0]}")
print(f"Filas afectadas: {test_data.index[test_data["Weight (kg)"].isna()]}")
print(f"Filas con HDL nula: {nulos_HDL.shape[0]}")
print(f"Filas afectadas: {test_data.index[test_data["HDL (mg/dL)"].isna()]}")
print(f"Filas con BMI nula: {nulos_BMI.shape[0]}")
print(f"Filas afectadas: {test_data.index[test_data["BMI"].isna()]}")
print(f"Filas con Height (m) nula: {nulos_height_m.shape[0]}")
print(f"Filas afectadas: {test_data.index[test_data["Height (m)"].isna()]}")
print(f"Filas con Abdominal Circum nula: {nulos_abdmomin.shape[0]}")
print(f"Filas afectadas: {test_data.index[test_data["Abdominal Circumference (cm)"].isna()]}")
print(f"Filas con height (cm) nula: {nulos_height_cm.shape[0]}")
print(f"Filas afectadas: {test_data.index[test_data["Height (cm)"].isna()]}")
print(f"Filas con Waist-height ratio nula: {nulos_w_h_ratio.shape[0]}")
print(f"Filas afectadas: {test_data.index[test_data["Waist-to-Height Ratio"].isna()]}")
print(f"Filas con Total Cholest nula: {nulos_colest.shape[0]}")
print(f"Filas afectadas: {test_data.index[test_data["Total Cholesterol (mg/dL)"].isna()]}")
print(f"Filas con Systolic nula: {nulos_systolic.shape[0]}")
print(f"Filas afectadas: {test_data.index[test_data["Systolic BP"].isna()]}")
print(f"Filas con Blood Sugar nula: {nulos_blood_sugar.shape[0]}")
print(f"Filas afectadas: {test_data.index[test_data["Fasting Blood Sugar (mg/dL)"].isna()]}")
print(f"Filas con Estimated LDL nula: {nulos_estimated_LDL.shape[0]}")
print(f"Filas afectadas: {test_data.index[test_data["Estimated LDL (mg/dL)"].isna()]}")

Filas con Age nula: 11
Filas afectadas: Index([16, 37, 47, 70, 90, 130, 137, 141, 155, 169, 171], dtype='int64')
Filas con Weight nula: 10
Filas afectadas: Index([10, 19, 105, 112, 149, 165, 169, 170, 174, 180], dtype='int64')
Filas con HDL nula: 8
Filas afectadas: Index([19, 97, 124, 135, 157, 166, 182, 191], dtype='int64')
Filas con BMI nula: 8
Filas afectadas: Index([12, 15, 104, 105, 114, 123, 167, 192], dtype='int64')
Filas con Height (m) nula: 7
Filas afectadas: Index([18, 43, 64, 67, 82, 114, 173], dtype='int64')
Filas con Abdominal Circum nula: 7
Filas afectadas: Index([11, 30, 46, 106, 148, 160, 187], dtype='int64')
Filas con height (cm) nula: 7
Filas afectadas: Index([14, 46, 52, 125, 138, 177, 186], dtype='int64')
Filas con Waist-height ratio nula: 6
Filas afectadas: Index([27, 34, 71, 96, 121, 150], dtype='int64')
Filas con Total Cholest nula: 6
Filas afectadas: Index([2, 28, 93, 125, 143, 166], dtype='int64')
Filas con Systolic nula: 6
Filas afectadas: Index([40, 81, 99, 1

## Preparación de datos

## Construcción del modelo de regresión lineal

## Evaluación cuantitativa

## Evaluación cualitativa

## Uso del modelo de regresión

## Respuestas a las preguntas del apartado Análisis de resultados

Una vez construido los modelos, deberías estar en capacidad de responder estas preguntas:

- ¿Cuál fue el valor de los diferentes coeficientes obtenidos en el mejor modelo?

- A partir de la tabla comparativa, ¿cuál modelo ofrece el mejor rendimiento sobre el conjunto test? ¿Qué interpretación puedes darles a los valores obtenidos sobre las métricas de rendimiento?

- ¿Cuáles variables fueron seleccionadas con el modelo seleccionado? A partir de estas, ¿qué interpretación de cara al problema puedes dar? Reflexiona sobre cómo este nuevo conocimiento podría ayudar a tomar decisiones en el contexto del problema.

- A partir del contexto y los datos compartidos, ¿cómo representar la regresión lineal de forma matemática? Indique el método utilizado y el proceso para resolverlo.

- En el ciclo de machine learning ¿Qué tipos de sesgo podría afectar los resultados y por qué? Describe dos tipos de sesgo.